include/adaflo/sharp_interface_util.h

```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
 
-  for (unsigned int i = 0; i < cells.size(); ++i)
-    {
-      typename DoFHandler<dim>::active_cell_iterator cell = {
-        &dof_handler.get_triangulation(), cells[i].first, cells[i].second, &dof_handler};
+        fe_eval.reinit(cell);
 
-      typename DoFHandler<dim>::active_cell_iterator cell_dim = {
-        &dof_handler.get_triangulation(),
-        cells[i].first,
-        cells[i].second,
-        &dof_handler_dim};
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_points.push_back(fe_eval.quadrature_point(q));
+      }
 
-      local_dof_indices.resize(cell->get_fe().n_dofs_per_cell());
-      buffer.resize(cell->get_fe().n_dofs_per_cell());
-      buffer_dim.resize(cell->get_fe().n_dofs_per_cell() * dim);
+    return integration_points;
+  }();
 
-      cell->get_dof_indices(local_dof_indices);
+  Utilities::MPI::RemotePointEvaluation<spacedim, spacedim> eval;
+  eval.reinit(integration_points, dof_handler.get_triangulation(), mapping);
 
-      const unsigned int n_points = ptrs[i + 1] - ptrs[i];
+  const auto integration_values = [&]() {
+    std::vector<T> integration_values;
 
-      const ArrayView<const Point<dim>> unit_points(points.data() + ptrs[i], n_points);
-      const ArrayView<const double>     JxW(weights.data() + ptrs[i], n_points);
+    FE_Nothing<dim, spacedim> dummy;
 
-      // gather curvature
-      constraints.get_dof_values(curvature_solution,
-                                 local_dof_indices.begin(),
-                                 buffer.begin(),
-                                 buffer.end());
+    FEValues<dim, spacedim> fe_eval(surface_mapping,
+                                    dummy,
+                                    surface_quadrature,
+                                    update_JxW_values);
 
-      // evaluate curvature
-      phi_curvature.evaluate(cell,
-                             unit_points,
-                             make_array_view(buffer),
-                             EvaluationFlags::values);
+    for (const auto &cell : surface_mesh.active_cell_iterators())
+      {
+        if (cell->is_locally_owned() == false)
+          continue;
 
-      // gather normal
-      for (int i = 0; i < dim; ++i)
+        fe_eval.reinit(cell);
+
+        for (const auto q : fe_eval.quadrature_point_indices())
+          integration_values.push_back(fe_eval.JxW(q));
+      }
+
+    return integration_values;
+  }();
+
+  const auto integration_function = [&](const auto &values, const auto &cell_data) {
+    AffineConstraints<double> constraints; // TODO: use the right ones
+
+    FEPointEvaluation<1, spacedim> phi_curvature(mapping, dof_handler.get_fe());
+
+    FESystem<spacedim>                    fe_dim(dof_handler.get_fe(), spacedim);
+    FEPointEvaluation<spacedim, spacedim> phi_normal(mapping, fe_dim);
+    FEPointEvaluation<spacedim, spacedim> phi_force(mapping, dof_handler_dim.get_fe());
+
+    std::vector<double>                  buffer;
```

Comment:
Question: Do you use this as
```suggestion
    std::vector<double>                  curvature_dof_values;
```

Reply:
The problem here is that these vectors are reused: for block vectors buffer is also used to read the components of the normal vector...